File: mergesort/mergesort.py
import math

################################################
## This function divides an array and merges it
## params : array, the first index, last index
## return : sorted array
################################################

def divideArray(arrayList, start, length):

	## check if the first index > last index
    if start < length:
        arraylength = len(arrayList)
        #print("array length", arraylength)
        middle = math.floor(arraylength // 2)
        #print("middle ", middle)
        array1 = arrayList[:middle] # get the first half / left side of the array by slicing the array
        array2 = arrayList[middle:] # get the second half or right side of the array by slicing the array
        divideArray(array1,start, middle) # recursion used to continually devide the array to 2 single arrray
        divideArray(array2,middle + 1, arraylength) # recursion used to continually devide the array to 2 single arrray
        mergesort(arrayList, array1, array2) #// return sorted array. See the recursive function used
		# if you do not know a bit of how merge sort works but you are unsure of the coding, you can add print statements on each line to view the steps


def mergesort(arrayList, array1, array2):
    i = 0
    j = 0
    k = 0

    #print("Original array is ", arrayList)
    #print("Left array is ", array1)
    #print("Right array is ", array2)

	# check if the length of both sliced array are > n ( n = 0, 1, 2...... array.length)
    while i < len(array1) and j < len(array2):
        if array1[i] <= array2[j]:
            arrayList[k] = array1[i] #swap the current value with the smallest value
            i = i + 1
            #print("Unsorted - 1 ", arrayList)
        else:
            arrayList[k] = array2[j] #swap the current value with the smallest value
            j = j + 1
            #print("Unsorted - 2 ", arrayList)
        k = k + 1
        #print("Unsorted ",arrayList)

	# check if the length of both sliced array are > n ( n = 0, 1, 2...... array.length)
    while i < len(array1):
        arrayList[k] = array1[i]
        i = i + 1
        k = k + 1

    while j < len(array2):
        arrayList[k] = array2[j]
        j = j + 1
        k = k + 1
    #print(arrayList)

arrll = [10, 14, 19, 26, 27, 31, 33, 35, 42, 44, 0]
divideArray(arrll, 0, len(arrll))
print(arrll)

File: mergesort/mergesort.js
/**
** This function divides an array and merges it
** params : array
** return : sorted array
**/
function divideArray(arrayList)
{
	/**
	** is the array length greater than 1?
	*/
    if(arrayList.length > 1) 
    {
        const middle = Math.floor(arrayList.length / 2) // get the floor value of the array after division
        const arr1 = arrayList.slice(0, middle) // get the first half / left side of the array
        const arr2 = arrayList.slice(middle) // get the second half or right side of the array

		// output the array
        console.log(arr1)
        console.log(arr2)

        return mergesort(divideArray(arr1), divideArray(arr2)) // return sorted array. See the recursive function used
		// if you do not know a bit of how merge sort works but you are unsure of the coding, you can add print statements on each line to view the steps
    }

    return arrayList
}


function mergesort(array1, array2)
{
    let i = 0
    let j = 0

    let arrayList = [] //initialise an empty array or you could parse the original array

	// check if the length of both sliced array are > n ( n = 0, 1, 2...... array.length)
    while(i < array1.length && j < array2.length)
    {
        if(array1[i] < array2[j]) // compare the values
        {
            arrayList.push(array1[i]) // push the smallest value
            i = i + 1
        }
        else
        {
            arrayList.push(array2[j]) // push the smallest value
            j = j + 1
        }
    }

    return arrayList.concat(array1.slice(i)).concat(array2.slice(j)) // merge the array and return it
}

let arrayL = [10, 14, 19, 26, 27, 31, 33, 35, 42, 44, 0]

console.log(divideArray(arrayL))